salt/auth: Add groups support for bearer auth

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'tests'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

Currently we have no permissions when using bearer token auth

**Summary**:

add a `groups` function to authenticate bearer auth using the same approach as
the basic auth for the moment
Also update test accordingly

TeddyAndrieux committed:
salt/auth: Add groups support for bearer auth

Currently we have no permissions when using bearer token auth, add a
groups function to authenticate bearer auth using the same approach as
the basic auth for the moment

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented on tests/post/features/salt_api.feature:
```diff
@@ -12,7 +12,10 @@ Feature: SaltAPI
+        Then we can ping all minions
+        And we can invoke '[".*"]' on '*'
+        And we have '@wheel' perms
+        And we have '@runner' perms
```
Uh, really? Should it?

gdemonet replied:
Because I don't think the Storage Operator needs to have `*` perms on Nodes

TeddyAndrieux replied:
> Should it?

I think no, but since we use Bearer token for storage-operator ... it is.
Since we have no "roles" we can only have every perms or None for the moment

> Because I don't think the Storage Operator needs to have `*` perms on Nodes

I agree

slaperche-scality replied:
Maybe related to https://github.com/scality/metalk8s/issues/1528

TeddyAndrieux replied:
Hmm yes true we have some rules in master config maybe something can be done.
Honnestly I don't really searched how our auth module work, just did a quick fix to have `groups` for Bearer auth.

gdemonet replied:
In master config, we have two "groups". One for node-admins, another for storage-operator. The SA for the Operator shouldn't have permissions that make it end up in the node-admins group.

> Maybe related to #1528

No, that one would be nice as well, but that's about reducing the permissions of the storage-operator group. Here, this test effectively confirms that we have the node-admins group, and that shouldn't be the case.

NicolasT replied:
Note: `storage-operator` is not a *group*, it's a SaltAPI *user*. Only `node-admins@` is a *group*.

TeddyAndrieux replied:
Ok just investigate, it's not really easy just retrieve the id_token from dex. (if you have idea do not hesitate)

So several choice for this PR:
- We test with `storage-operator` token as this specific user has the right perms (see https://github.com/scality/metalk8s/pull/2068/files#r351360140)
- We create a new service account just to do this test
- We not test in this PR as this full auth process will be tested by cypress test in #2042

gdemonet replied:
I'm getting confused. Why are we talking about Dex? Anything that runs in the cluster can rely on ServiceAccounts, and that's what `storage-operator` uses (or should use) as well.

See:
https://github.com/scality/metalk8s/blob/development/2.4/storage-operator/deploy/role.yaml
https://github.com/scality/metalk8s/blob/development/2.4/storage-operator/deploy/service_account.yaml

TeddyAndrieux replied:
Since we use dex we can authenticate without service account using directly dex that will give you a token for a user which is bind in `dex-administrator` cluster role binding. But I didn't found any simple solution to retrieve this token (and also it's not really part of salt-api stuff)

I know it's a service account that my point, 
So we need to create a new service account for this test or use the storage-operator one.

gdemonet replied:
> we need to create a new service account for this test or use the storage-operator one.

I don't understand the problem. In the Gherkin, we have:

```
When we login to SaltAPI with the ServiceAccount 'storage-operator'
```

I would then expect that's already what the test does. We shouldn't try to authenticate without a SA from those tests (yet, that's something to do, for sure, but will take some time to implement, with the whole OAuth dance).

TeddyAndrieux replied:
Yes but storage operator is a special case as this user is define directly in the salt-master it may make sense to have a new SA to test the groups function for bearer auth

NicolasT replied:
Ah, right, I think I get it now.

So indeed, there is currently no test to check whether *with a given token* for a user that's *not* `storage-operator` but *does* have the rights to do `*` on `Nodes` would get assigned to the 'right' group (i.e., `node-admins`).

So, yes, maybe creating a SA which then gets a `ClusterRoleBinding` to `system:masters`, then check whether we get the expected access to SaltAPI when using this SA's token would work.

No need for Dex or whatnot, so could (should?) be part of this PR.

gdemonet commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -114,14 +114,54 @@ def _groups_basic(kubeconfig, username, token):
+                ),
+            ),
+        ),
+    )
```
Could be worth factorizing with the code in `_groups_basic`

TeddyAndrieux replied:
Since this one should change in the futur (and I hope before 2.5.0 release), to handle different roles I think it's not really needed

gdemonet replied:
Disagree. The "node-admins" group for Salt will still exist I guess, and even if the way to know if a user is in that group changes in the future, there's common logic between Basic and Bearer tokens.

NicolasT replied:
Agree with @gdemonet. The mapping of K8s rights to Salt groups is intricate, and should be *the same* whether you're using Basic auth, Bearer tokens or whatever. The only difference is the mangling of the `kubeconfig` object (which is an ugly hack anyway).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

NicolasT commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -114,14 +114,54 @@ def _groups_basic(kubeconfig, username, token):
+    kubeconfig.api_key_prefix = {
+        'authorization': 'Bearer',
+    }
+    kubeconfig.username = username
```
This shouldn't be set I believe, this information is contained in the token.

NicolasT commented on tests/post/features/salt_api.feature:
```diff
@@ -12,7 +12,10 @@ Feature: SaltAPI
         Given the Kubernetes API is available
         When we login to SaltAPI with the ServiceAccount 'storage-operator'
-        Then we can invoke '["disk.dump", "state.sls"]' on '*'
+        Then we can ping all minions
```
Basically, the fact that this test requires changes shows that there's an issue somewhere in the implementation. This test should remain as-is.

TeddyAndrieux replied:
Btw this test do not require change but since `storage-opeator` is able to do a `*` on `nodes` he also have this perms.

Basically storage-operator perms in this PR are:
```
        "perms": [
            {
                "*": [
                    "disk.dump", 
                    "state.sls"
                ]
            }, 
            "@jobs", 
            {
                "*": [
                    ".*"
                ]
            }, 
            "@wheel", 
            "@runner", 
            "@jobs"
        ], 
```

gdemonet replied:
But it shouldn't have those perms?!? https://github.com/scality/metalk8s/blob/development/2.4/storage-operator/deploy/role.yaml

TeddyAndrieux replied:
As the cluster role say that he can do `*` on nodes the check we do for groups is approved so we add node-admin groups thats why but if we dont want that the storage operator can do this then we need to either change the check from groups function or changes the rule of his cluster role

gdemonet replied:
Oh right, I read the Role too fast. Then that would be a bug I guess, we shouldn't give that many permissions to `storage-operator`.

gdemonet replied:
From [this issue](https://github.com/operator-framework/operator-sdk/issues/1226), it seems we should not have kept the default ClusterRole that gets created when the project is initialized with `operator-sdk new`. Since `operator-sdk generate k8s` doesn't generate this file, it would make sense to update it to match our needs and commit it.

gdemonet replied:
See #2084

TeddyAndrieux replied:
Then it's not part of this PR, because we may want to fix it also in 2.4

gdemonet commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -58,6 +58,30 @@ def _check_k8s_creds(kubeconfig, token):
         raise
 
 
+def _get_groups(kubeconfig):
```
Works for now, since we only support one group, but I would have imagined something that checks whether to add the 'node-admins' group and only return this information, and be able to call a bunch of such single-focus checks to compute the list/set of groups for a token.

TeddyAndrieux replied:
Ok make sense i will update it

TeddyAndrieux replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

TeddyAndrieux committed:
tests: Add Admin ServiceAccount fixture

Add a fixture to create an Admin ServiceAccount for test purpose.
This fixture create a ServiceAccount and a ClusterRoleBinding to bind
this new ServiceAccount to `cluster-admin` ClusterRole

TeddyAndrieux committed:
tests: Add salt-api test for Bearer token of Admin SA

Add test that a ServiceAccount bind to `cluster-admin` ClusterRole is
able to everything with salt-api

TeddyAndrieux commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
